internal-v2.nix

```diff
@@ -0,0 +1,556 @@
+  readPackageLikeFile = file:
+    assert (builtins.typeOf file != "path" && builtins.typeOf file != "string") ->
+      throw "file ${toString file} must be a path or string";
+    let
+      content = builtins.readFile file;
+      json = builtins.fromJSON content;
+    in
+    assert
+    builtins.typeOf json != "set" ->
+    throw "The NPM lockfile must be a valid JSON object";
+    # if a lockfile doesn't declare dependencies ensure that we have an empty
+    # set. This makes the consuming code eaiser.
+    if json ? dependencies then json else json // { dependencies = { }; };
+
+  # Description: Patch a lockfile v2 package entry. It'll replace the
+  # URL stored in the integrity field with nix store path.
+  # spec :: {version :: String, resolved :: String, integrity :: String }.
+  # Type: { version :: String, resolved :: String, integrity :: String }
+  patchPackage = sourceOptions@{ sourceHashFunc, ... }: raw_name: spec:
+    assert (builtins.typeOf raw_name != "string") ->
+      throw "Name of dependency ${toString raw_name} must be a string";
+    assert !(spec ? resolved) ->
+      throw "Missing resolved field for dependency ${toString raw_name}";
+    assert !(spec ? version) ->
+      throw "Missing version field for dependency ${toString raw_name}";
+    let
+      name = genericPackageName raw_name;
+      defaultedIntegrity = if spec ? integrity then spec.integrity else null;
+      # Relaxing dependencies version bounds: it could be a GitHub
+      # ref, forcing NPM to checkout the remote repo to get the actual
+      # version.
+      # We already pinned everything through the "resolved", we can
+      # relax those.
+      patchDependencies = deps: lib.mapAttrs (_n: dep: if isGitHubRef dep || isGitHubRefWithoutRev dep then "*" else dep) deps;
+      patchedResolved =
+        if (!isGitHubRef spec.resolved)
+        then makeUrlSource sourceOptions name spec.version spec.resolved defaultedIntegrity
+        else
+          let
+            ghRef = parseGitHubRef spec.resolved;
+            ghTgz = buildTgzFromGitHub {
+              inherit name sourceOptions;
+              inherit (ghRef) org repo rev;
+              ref = ghRef.rev;
+              hash = sourceHashFunc { type = "github"; value = { inherit (ghRef) org repo rev; }; };
+            };
+          in
+          {
+            resolved = "file://" + (toString ghTgz);
+            integrity = null;
+          };
+    in
+    spec // {
+      inherit (patchedResolved) resolved integrity;
+    } // lib.optionalAttrs (spec ? dependencies) { dependencies = (patchDependencies spec.dependencies); };
+
+  genericPackageName = name:
+    (lib.last (lib.strings.splitString "node_modules/" name));
+
+  # Description: Takes a parsed lockfile and returns the patched version as an attribute set
+  # Type: { sourceHashFunc :: Fn } -> parsedLockedFile :: Set -> { result :: Set, integrityUpdates :: List { path, file } }
+  patchLockfile = sourceOptions: content:
```

Comment:
Some node modules seem to provides package.json with dependencies with "*" version in v2 lockfile. I managed to work around this issue by finding the latest package version to replace those. I will try to find a repro case.

Reply:
Do you mean the `packages.$name.resolved` field end up with a "*" value? :scream: If that's the case, I did not hit that one yet. I'm not even sure we can do something here, I'd consider that as a npm bug we should fix upstream.

If you're talking about the `packages.$name.dependencies.$name`, I'd argue it's not an issue for us. We exclusively use the `resolved` field to remotely download packages in the v2 codebase.

Reply:
Finally the issue is when package.json have "latest" as version specs. Not sure if this is a "dependencies" only issue. ex. https://github.com/compodoc/live-server/blob/master/package.json